Add more HtmlUtils APIs

andy-armstrong opened the pull request.

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

andy-armstrong commented:
@robrap Here's my first attempted update to the APIs. I'm going to move on to update my other PR now, but I'd welcome your feedback here too (or you can wait to see the APIs in action).

robrap commented:
@andy-armstrong Looks reasonable.  I need to get back to my docs.  The real question is how this play out in your other PRs.  Also, maybe @cahrens can help review the comments and make sure they are still correct?

cahrens commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,75 @@
-                return text.toString();
+        ensureHtml = function(html) {
+            if (html instanceof HtmlSnippet) {
+                return html.toString();
```
I don't understand the inconsistent return values of this method. In this case (html already an HtmlSnippet), this method returns a String. But in the case of a String being passed in, an HTMLSnippet is returned.

andy-armstrong replied:
My bad. This should be returning an HtmlSnippet in both cases now.

andy-armstrong replied:
I've fixed it and added a test.

cahrens commented on src/js/utils/html-utils.js:
```diff
@@ -109,15 +142,60 @@
+        /**
+         * A wrapper for $.html that safely escapes the provided HTML.
+         *
+         * If the HTML is provided as an HTML snippet then it is set directly
```
I think you should include this explanation in the other two methods below as well.

andy-armstrong replied:
Fixed.

andy-armstrong commented:
@robrap @cahrens This is ready for re-review. 

FYI @dan-f @AlasdairSwan

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,78 @@
 
         /**
-         * Returns an HTML string that is appropriately escaped.
+         * Ensures that the specified HTML is encoded as an HtmlSnippet.
```
Ensures that the provided text is properly HTML-escaped.

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,78 @@
+         * Ensures that the specified HTML is encoded as an HtmlSnippet.
          *
-         * Note: if the string provided is flagged as being an HTML snippet
+         * If the HTML is provided as a string, then it is HTML-escaped and
```
If a plain text string is provided, then it will be HTML-escaped and...

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,78 @@
-         * @param {string} text A string that is either plan text, or is
-         * flagged as being an HTML snippet.
-         * @returns {string} A safely HTML-escaped string.
+         * @param {(string|HtmlSnippet)} html HTML provided either as
```
```
* @param {(string|HtmlSnippet)} html Either a plain text string or an HTML snippet.
```

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,78 @@
+            if (!(html instanceof HtmlSnippet)) {
+                throw 'HtmlUtils.element received HTML that was not an HtmlSnippet: ' + html;
+            }
+            return $(ensureHtml(html).toString());
```
You could reverse the order of these and have an if/else, where the if doesn't require the "not".  Additionally, you don't need ensureHtml() if you are already checking that this must be an HtmlSnippet.

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,78 @@
+        /**
          * Returns an HTML snippet by interpolating the provided parameters.
          *
          * The HTML text is provided as a tokenized format string where parameters
```
The text is provided...

It's probably plain text that is being provided, not HTML.

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -76,19 +113,39 @@
          *
          * @param {string} formatString The string to be interpolated.
          * @param {Object} parameters An optional set of parameters to the template.
```
The word "template" seems odd here.  Instead, maybe "...parameters for interpolation."

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -107,17 +164,71 @@
+         * so it is first escaped before being used.
+         *
+         * @param {element} element The element or elements to be updated.
+         * @param {(string|HtmlSnippet)} html The desired HTML, either as a string
```
... either as a plain text string or an HTML snippet.

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -107,17 +164,71 @@
+         * so it is first escaped before being used.
+         *
+         * @param {element} element The element or elements to be updated.
+         * @param {(string|HtmlSnippet)} html The desired HTML, either as a string
```
Same as above.

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,41 +6,78 @@
+         * @param {HtmlSnippet} html The desired HTML provided as an HtmlSnippet.
+         * @returns {element} A new JQuery element.
+         */
+        element = function(html) {
```
Where and how is this method used?

andy-armstrong replied:
I've decided to remove this method for now so we can get this PR merged.

I was thinking we'd need it for calls to ``appendTo`` and ``prependTo`` where the caller has to create a new element from some HTML before attaching it. Here's one example:

https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/progress_graph.js#L9

I'll revisit it when I have some more concrete examples where the existing methods won't be enough.

robrap commented on src/js/utils/specs/html-utils-spec.js:
```diff
@@ -91,20 +114,131 @@ define(
+                        '<a href="world">world</a>'
+                    ],
+                    'can join a mixture of strings and HTML snippets': [
+                        ['Hello, ', HtmlUtils.HTML('<a href="world">world</a>')],
```
"Hello" should be "Rock & Roll" if you really want this to test a mix.

cahrens commented:
Sorry, I don't have time to review today due to meetings and the fact that I have to leave at 2:45.

andy-armstrong commented:
@robrap I've addressed your feedback if you could do a quick re-review.

@AlasdairSwan would you be able to review this as @cahrens won't be able to today.

AlasdairSwan commented on src/js/utils/html-utils.js:
```diff
@@ -75,20 +95,40 @@
+        joinHtml = function() {
+            var html = '',
+                i;
+            for (i = 0; i < arguments.length; i++) {
```
Nitpick: the length should be assigned to a variable instead of calculating length on each iteration.

AlasdairSwan commented:
Just one minor nitpick @andy-armstrong . :+1:

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -6,44 +6,64 @@
-         * Helper function to return a string flagged as an HTML snippet.
+         * Creates an HTML snippet.
+         *
+         * The intention of an HTML snippet is to capture a string that
```
The intention of an HTML snippet is to communicate that the string it represents contains HTML that has been safely escaped as necessary.

andy-armstrong replied:
Fixed.

robrap commented on src/js/utils/html-utils.js:
```diff
@@ -107,17 +147,70 @@
+         * A wrapper for $.append that safely escapes the provided HTML.
+         *
+         * If the HTML is provided as an HTML snippet then it is used directly.
+         * If the value is a string then it is assumed to be unescaped and
```
Are there ever real reasons to append a plain text string that gets escaped, or is this just a safety feature?  Same question for `setHtml()` above.  If it is just a safety feature, we might want to adjust the text to make that clear.

andy-armstrong replied:
See my comment on the other PR: https://github.com/edx/edx-platform/pull/11856/files#r57047568

IMO it is useful for ``setHtml`` and arguable for ``append``. I'd like to keep it for consistency and to enforce safety as you point out.

robrap replied:
@andy-armstrong I was confirming and thinking the docs might want to make that clear, but I think it is fine and we'll see what happens.  Thanks.

robrap commented:
@andy-armstrong Minor comment and minor question.  I'll help you get to thumb as soon as you are available.

robrap commented:
:+1:   Thanks @andy-armstrong
